apis/openstack-swift/src/main/java/org/jclouds/openstack/swift/v1/blobstore/RegionScopedSwiftBlobStore.java

```diff
@@ -582,19 +589,58 @@ public long countBlobs(String containerName, ListContainerOptions options) {
    // copied from BaseBlobStore
    @Beta
    protected String putMultipartBlob(String container, Blob blob, PutOptions overrides) {
-      List<MultipartPart> parts = Lists.newArrayList();
+      ConcurrentLinkedQueue<MultipartPart> parts = new ConcurrentLinkedQueue<MultipartPart>();
+
       long contentLength = checkNotNull(blob.getMetadata().getContentMetadata().getContentLength(),
             "must provide content-length to use multi-part upload");
       MultipartUploadSlicingAlgorithm algorithm = new MultipartUploadSlicingAlgorithm(
             getMinimumMultipartPartSize(), getMaximumMultipartPartSize(), getMaximumNumberOfParts());
       long partSize = algorithm.calculateChunkSize(contentLength);
       MultipartUpload mpu = initiateMultipartUpload(container, blob.getMetadata(), partSize, overrides);
       int partNumber = 1;
+
+      int numThreads = 16;
+      if (System.getProperty("jclouds.mpu.parallel.degree") != null) {
+         numThreads = Integer.getInteger(System.getProperty("jclouds.mpu.parallel.degree"));
+      }
+
+      BlockingQueue<Runnable> blockingQueue = new ArrayBlockingQueue<Runnable>(numThreads);
+      RejectedExecutionHandler rejectedExecutionHandler = new ThreadPoolExecutor.CallerRunsPolicy();
+       ExecutorService executor =  new ThreadPoolExecutor(numThreads, numThreads,
+            0L, TimeUnit.MILLISECONDS, blockingQueue, rejectedExecutionHandler);
+
       for (Payload payload : slicer.slice(blob.getPayload(), partSize)) {
-         MultipartPart part = uploadMultipartPart(mpu, partNumber, payload);
-         parts.add(part);
          ++partNumber;
+         BlobUploader b =
+               new BlobUploader(mpu, partNumber, payload, parts);
+         executor.execute(b);
+      }
+      executor.shutdown();
+      try {
+         executor.awaitTermination(Integer.MAX_VALUE, TimeUnit.DAYS);
+      } catch (InterruptedException e) {
+         e.printStackTrace();
+      }
+
+      return completeMultipartUpload(mpu, new ArrayList( parts ));
+   }
+
+   private class BlobUploader implements Runnable {
```

Comment:
Making it static means uploadMultipartPart in BlobStore will have be made static, otherwise it's referencing a dynamic context from a static one. But if I do that it would mean changes in other providers and so forth. There might be other problems with that, too.